design/users/_auth.json

```diff
@@ -1 +1 @@
f(userCtx.name) !== -1) {\n          return true; // database admin\n        }\n      }\n\n      // see if the user a database admin specified by role\n      if(secObj && secObj.admins && secObj.admins.roles) {\n        var db_roles = secObj.admins.roles;\n        for(var idx = 0; idx < userCtx.roles.length; idx++) {\n          var user_role = userCtx.roles[idx];\n          if(db_roles.indexOf(user_role) !== -1) {\n            return true; // role matches!\n          }\n        }\n      }\n\n      return false; // default to no admin\n    }\n\n    var filterDifferentRoles = function(array) {\n      return function(role) { return array.indexOf(role) === -1 };\n    }\n\n    var combineArrays = function(array1, array2) {\n      return array1.concat(array2.filter(filterDifferentRoles(array1)));\n    }\n\n    var onlyLearnerChange = function(oldRoles, newRoles) {\n      var removedRoles = oldRoles.filter(filterDifferentRoles(newRoles));\n      var addedRoles = newRoles.filter(filterDifferentRoles(oldRoles));\n      var changedRoles = combineArrays(removedRoles, addedRoles);\n      return changedRoles.length === 1 && changedRoles[0] === 'learner';\n    }\n\n    if (!is_server_or_database_admin(userCtx, secObj)) {\n      var openLearnerRole = secObj.admins.roles.indexOf('openlearner') > -1;\n      if (oldDoc) { // validate non-admin updates\n        if (userCtx.name !== newDoc.name) {\n          throw({\n            forbidden: 'You may only update your own user document.'\n          });\n        }\n        // validate role updates\n        var oldRoles = (oldDoc.roles || []).sort();\n        var newRoles = newDoc.roles.sort();\n\n        if (openLearnerRole && onlyLearnerChange(oldRoles, newRoles)) {\n          return;\n        }\n\n        if (oldRoles.length !== newRoles.length) {\n          throw({forbidden: 'Only _admin may edit roles'});\n        }\n\n        for (var i = 0; i < oldRoles.length; i++) {\n          if (oldRoles[i] !== newRoles[i]) {\n            throw({forbidden: 'Only _admin may edit roles'});\n          }\n        }\n      } else if (newDoc.roles.length > 0) {\n        if (!openLearnerRole || newDoc.roles.length > 1 || newDoc.roles[0] !== 'learner') {\n          throw({forbidden: 'Only _admin may set roles'});\n        }\n      }\n    }\n\n    // no system roles in users db\n    for (var i = 0; i < newDoc.roles.length; i++) {\n      if (newDoc.roles[i][0] === '_') {\n        throw({\n          forbidden:\n          'No system roles (starting with underscore) in users db.'\n        });\n      }\n    }\n\n    // no system names as names\n    if (newDoc.name[0] === '_') {\n      throw({forbidden: 'Username may not start with underscore.'});\n    }\n\n    var badUserNameChars = [':'];\n\n    for (var i = 0; i < badUserNameChars.length; i++) {\n      if (newDoc.name.indexOf(badUserNameChars[i]) >= 0) {\n        throw({forbidden: 'Character `' + badUserNameChars[i] +\n            '` is not allowed in usernames.'});\n      }\n    }\n  }"}
```

Comment:
@wanglihe2010 This file should not be included in the branch.  You can remove it from the branch with
```
git checkout origin/master design/users/_auth.json
```
In the future, be careful with `git add` because this file sometimes changes after `vagrant up`

Reply:
Hi, Paul, I tried this command, but the file is still in the commit stage. How to remove it from previous commit?

Reply:
Since this file is different on master vs your branch, you needed to add & commit the file again after running the command above to revert the change.  I'm pushing this change now.